feat(pip_repository): Support pip parse cycles

arrdem opened the pull request:
This patch reworks the `pip_repository` machinery to allow users to manually annotate groups of libraries which form packaging cycles in PyPi and must be simultaneously installed.

The strategy here is to transform any dependencies `A` and `B` which have dependencies and are mutually dependent

```mermaid
graph LR;
    A-->B;
    A-->D;
    A-->E;
    B-->A;
    B-->F;
    B-->G;
```

into a new "dependency group" `C` which has `A*` and `B*`  as dependencies, defined as `A` and `B` less any direct dependencies which are members of the group. This is viable _for python_ because Python files just need to be emplaced into a runfiles directory for the interpreter. We don't actually have a true hard dependency between the build definition of `A` requiring the build product `B` be available which requires that the build product of `A` be available.

```mermaid
graph LR
     C-->A*;
     A*-->D;
     A*-->E;
     C-->B*;
     B*-->F;
     B*-->G;
```
This gets us most of the way there, as a user can now safely write `requirement("A")` and we can provide them with `C`, which has the desired effect of pulling in `A`, `B` and their respective transitives.

There is one remaining problem - a user writing `deps = [requirement("A"), requirement("B")]` will take a double direct dependency on `C`. So we need to insert a layer of indirection, generating `C_A` and `C_B` which serve only as unique aliases for `C` so that we can support the double dependency. Our final dependency graph then is as follows

```mermaid
graph LR
     C_A-->C;
     C_B-->C;
     C-->A*;
     A*-->D;
     A*-->E;
     C-->B*;
     B*-->F;
     B*-->G;
```

Addresses #1076, #1188

## To do
- [x] Get rebased
- [x] Get re-validated manually
- [x] Buildifier
- [x] Get CI happy
- [x] Update documentation
- [x] Update changelog

arrdem commented:
Hrm. Maybe I'm missing something here but it looks like bazel-under-bazel for the tests isn't happy with the two new templates. Not sure how to proceed with a fix.

arrdem commented:
Argh. Thanks @chrislovecnm, will try and get that fixed soonish.

arrdem commented:
Okay this needs some work to fit in with the bzlmod support. Will update when I've gotten that ironed out.

rickeylev commented:
@groodt

arrdem commented:
After dusting this off and starting some testing, the original tack of this patch (creating a single shared `c` library wrapping `a` and `b` to make them simultaneously depended without a mutual dependency) doesn't succeed in providing the same interface as the current `requirement()` machinery because `deps = [requirement("a"), requirement("b")]` is `deps = [requirement("c"), requirement("c")]` which is invalid.

With some additional head standing it may be possible to do something like generating a unique alias to `c` for `a` and `b` each, or generating n-many `c` equivalent library targets to mitigate this duplicate dependency problem.

Alternatively the `py_library` and `py_binary` macros could do deduplication but I doubt that's a change that would be of interest.

arrdem commented:
Okay - I think this is ready for a review. Includes test updates, passes the test updates. The bzlmod story for this is ... there isn't one yet. If there's a way to map this strategy onto bzlmod I'm all ears, I'm just not familiar with that machinery and opted to partition it and retain current `main` behavior.

chrislovecnm reviewed (changes requested):
I appreciate your work! We are having problems on windows with this PR.  Please take a look at CI - https://buildkite.com/bazel/rules-python-python/builds/4896#01882c9e-c77f-4596-bfbc-8ca5b5c2931a

chrislovecnm commented on examples/pip_parse_vendored/requirements.bzl:
```diff
@@ -7,47 +7,149 @@ from //:requirements.txt
+        return "@pip//:whl_" + _cluster_mappings[cname]
+    return "@pip_" + cname + "//:whl"
 
+# buildifier: disable=function-docstring
```
I don't think this is new code, but how was it failing buildifier?

arrdem replied:
`buildifier -lint -r .` was barking in `examples/pip_parse_vendored` (6.1.0 from upstream) without this last night while I was trying to avoid iterating on this using CI. I was surprised to be getting function docstring warnings without this because as you note this example originally didn't need these disables. Can run doing without it by CI and see what it says.

arrdem replied:
https://buildkite.com/bazel/rules-python-python/builds/4908#01883047-b108-4b99-b24d-d92a1a8d241b/118-119 recurred in CI when I removed these lines. Putting 'em back.

chrislovecnm commented on examples/pip_parse_vendored/requirements.bzl:
```diff
@@ -7,47 +7,149 @@ from //:requirements.txt
             **whl_config
         )
+
+    # And deal with requirement_clusters
```
Can you include more comments? Really spell out what "dealing" with requirement_clusters means.

chrislovecnm commented on python/pip_install/pip_repository.bzl:
```diff
@@ -431,14 +429,44 @@ def _pip_repository_impl(rctx):
     parsed_requirements_txt = parse_requirements(content)
 
-    packages = [(_clean_pkg_name(name), requirement) for name, requirement in parsed_requirements_txt.requirements]
+    # Apply name normalizations to the composite libs def once
```
Doc recommendations:

How and why are we normalizing this data?  What are the composite libs?

chrislovecnm commented on python/pip_install/pip_repository.bzl:
```diff
@@ -431,14 +429,44 @@ def _pip_repository_impl(rctx):
+    }
 
-    bzl_packages = sorted([name for name, _ in packages])
+    # Ditto for requirements defs
```
Change Ditto to something like, as in the code block about we are doing this.

chrislovecnm commented on python/pip_install/pip_repository.bzl:
```diff
@@ -775,4 +823,4 @@ def _repr_dict(items):
 
 def _format_dict(items):
-    return "{{{}}}".format(", ".join(sorted(['"{}": {}'.format(k, v) for k, v in items.items()])))
+    return "{{\n{}}}".format("".join(sorted(['    "{}": {},\n'.format(k, v) for k, v in items.items()])))
```
Can we use a platform independent newline?  It is a nit pick.

arrdem replied:
Would be happy to, not sure what that would be and not finding obvious documentation hits searching

arrdem replied:
Wound up just reverting this to avoid possible windows/bzlmod impact.

chrislovecnm commented on python/pip_install/tools/dependency_resolver/__init__.py:
```diff
@@ -11,4 +11,3 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-
```
Nit pick.  Please remove the file cleanup in this PR.  I would like to not have extra needed files in the commit.

chrislovecnm commented on python/pip_install/tools/wheel_installer/wheel_installer.py:
```diff
@@ -318,15 +322,21 @@ def _extract_wheel(
+    to_skip = {bazel.sanitise_name(it, "") for it in [whl.name] + skip_deps}
+    deps = {bazel.sanitise_name(it, "") for it in whl.dependencies(extras_requested)}
+    whl_deps = sorted(deps - to_skip)
+    print(
```
I have seen "print" statements through buidifier warnings.

chrislovecnm commented:
@arrdem please rebase.

chrislovecnm reviewed (changes requested):
First, thank you for the work.  Second, I need your thoughts about how we will do multiple pip support in bzlmod.  With these changes, you did you have knowledge that I don't have about this code.

My overall comments:

- I love to see over-documented code.  You can blame Tim Hockin for this. He taught me we cannot remember why we wrote code changes three months ago.
- Have we tested this with gazelle?  How does gazelle interact with this new code?  Hopefully, it ignores it, but we may want to document that gazelle does not support this and file an issue.

Kudos on the PR description.  That is awesome.

chrislovecnm commented:
Will this address https://github.com/bazelbuild/rules_python/issues/1188?

chrislovecnm commented:
> There is one remaining problem - a user writing deps = [requirement("A"), requirement("B")] will take a double direct dependency on C. So we need to insert a layer of indirection, generating C_A and C_B which serve only as shims to C so that we can support the double dependency.

If we are not addressing this problem, do we have an open issue?

chrislovecnm commented:
@aignas, do you mind taking a look?

arrdem commented:
> Will this address #1188?

I believe it will.



> > There is one remaining problem - a user writing deps = [requirement("A"), requirement("B")] will take a double direct dependency on C. So we need to insert a layer of indirection, generating C_A and C_B which serve only as shims to C so that we can support the double dependency.
> 
> If we are not addressing this problem, do we have an open issue?

I'll edit the PR description a bit to clarify. This problem was present in the original PR I submitted and was fixed yesterday. Note that [`requirement`](https://github.com/bazelbuild/rules_python/pull/1166/files#diff-45c1d09a54febf55b1b7c163b36d1940d9e3acfd61561682591f0ba0b5a1d516R29) is generating a reference to an [alias rule](https://github.com/bazelbuild/rules_python/pull/1166/files#diff-776a26dd44b340736e05824b49497b4534ea531698cf357c7dfff98266aa8365R14) to resolve this.

arrdem commented:
Tightened this up a bunch. Removed all unneeded changes that impacted the bzlmod machinery. Backed out some of the formatting changes. Eliminated the generated `lib.bzl` file. Updated all the code, comments and attrs to consistently refer to `requirement clusters`.

As far as I can determine at this point the bzlmod CI failures on windows are unrelated. I don't have a windows test environment available and I've eliminated all the parts of this changeset which impact bzlmod machinery. Guidance on getting to green there would be appreciated.

potiuk commented:
Hello here. I am  apache-airflow maintainer, and we used to have similar problem as other problems mentioned here: https://github.com/apache/airflow/issues/17795 

We "fixed" in the past by removing one part of the dependencies, but it starts causing us real issues, and we are going to bring dependency back, so without this PR bazel users will start to have again similar problems. We tried to workaround them but it causes more harm than good, so we are going to stop worrying about bazel (and we wil even explicitly state that bazel might have problems installing Airlfow (and we remove it when this PR is released). 

As mentioned in https://github.com/apache/airflow/issues/17795#issuecomment-904329243 by maintainer of `pip`, dependencies of packages in PyPI are not DAGs, and cycles are allowed, so technically it's Bazel that is not following the standard (that's why I am happy this PR is here).

rickeylev commented:
Thanks for the inline comments -- helps a lot.

re: bzlmod support: this should work with bzlmod; I don't see why not? In any case, we need to make sure it's possible with bzlmod because we'll be going bzlmod-only in a year or so.

re: requirements cluster key: This is basically an arbitrary, internall value, right? It just gets used as the internal cluster target name.

arrdem commented on python/pip_install/pip_repository_requirements.bzl.tmpl:
```diff
@@ -4,24 +4,42 @@
-    return "@%%NAME%%_" + _clean_name(name) + "//:pkg"
+def requirement(name, use_clusters = True):
+    cname = _clean_name(name)
+    if cname in _requirement_cluster_mapping and use_clusters:
+        return "@%%NAME%%//:_cluster_{}_lib_{}".format(_requirement_cluster_mapping[cname], cname)
+    else:
+        return "@%%NAME%%_" + cname + "//:pkg"
```
I'm pretty sure that's true, but I need to think through it and double check the wheel installer logic.

arrdem commented:
> re: bzlmod support: this should work with bzlmod; I don't see why not? In any case, we need to make sure it's possible with bzlmod because we'll be going bzlmod-only in a year or so.

Yes, this strategy (and probably all the associated machinery) should work under bzlmod, I'm just not familiar with how bzlmod works and we haven't adopted it yet so I haven't tried to wire it up here.

> re: requirements cluster key: This is basically an arbitrary, internal value, right? It just gets used as the internal cluster target name.

Correct. We could maybe do away with it entirely by accepting a list of clusters and generating ordinal identifiers internally, but I'm not sure that's better at all.

chrislovecnm commented:
@rickeylev @arrdem, the comment about bzlmod is that we are not doing a bzlmod integration test with the example that you using to test this.  Is adding a test to the use under /examples/bzlmod possible?

chrislovecnm commented:
After #1254 I think we are at a pretty stable place with pip and bzlmod.  I'm sure that we will find more to do, but those changes will probably cause some heartache with merging.  I'm not planning on more big rewrites of pip and bzlmod.

chrislovecnm commented:
@arrdem so the PR is in, and the interface for pip is pretty stable.  If you could merge in I would love to get this in.

aignas commented on python/pip_install/pip_repository_requirements.bzl.tmpl:
```diff
@@ -4,24 +4,42 @@
 
-def requirement(name):
-    return "@%%NAME%%_" + _clean_name(name) + "//:pkg"
+def requirement(name, use_clusters = True):
```
+1 for the users not needing to care. It seems that using this with `gazelle` could be problematic as `gazelle` would expect all of the available labels to be present as `@pip//foo` or `@pip_foo//:pkg` and the label returned by this `requirement` macro is neither. Maybe we could come up with a way to define an external repo here?

alexeagle commented:
I'm curious how this interacts with @philsc work to move fetches into the Bazel downloader. Will that also need to collapse cycles into a "cluster"?

philsc commented:
If the "cluster" approach is the desired approach, then yeah, I'll have to adopt it. Thanks for the heads up @alexeagle !

The way I solved this previously is to add annotation support for cutting dependencies. Circular dependencies really don't make sense and should get fixed upstream. I.e. our approach was to cut the dependency graph at a reasonable point and then file upstream tickets for them to fix it.

arrdem commented:
@shicholas the stopgap measure is to depend on rules_python via a `git_remote` with this patchset (or an equivalent) applied which is what my employer has been doing since this code reached feature completion in May. No promises on when I'll be able to revisit this, although my Q3 plan is full of Bazel stuff.

groodt commented:
I can tell you what we do at Canva: we patch wheels and / or BUILD files as necessary to fix broken upstream dependencies.

We often do try to work with upstream to fix their dependencies where possible. I personally don't think a clustering solution is really necessary here. There aren't that many places where there are dependencies between distributed packages and typically an end-user will just be taking a dependency on one of them at runtime. And for the other handlful of scenarios where it's not possible, the users just bring along the missing dependency in their targets.

The issue here isn't that python supports circular dependencies at runtime, its just that everything is expected to be installed into a flat dependency namespace and they don't really need to care about cycles because they aren't building a DAG. They're just downloading and installing a bunch of zip files into site-packages. Naturally, bazel can't do this, but we can easily handle it with simple patching of packages.

potiuk commented:
Just to mention a case that just patching packages is not a good solution because there are legitimate cases where it makes sense to have circular deps and patching packages will break the dependencies.

Example: we have airflow broken into 'apache-airflow' and 80+ providers. For example 'apache-airflow-providers-smtp'. The SMTP provider is a required dependency of 'apache-airflow'. But at some point we bump minimum version of supported airflow for our providers, so currently all our providers cannot be installed together with apache-airflow < 2.4.0 (API changed and they won't work). So SMTP provider has also the required dependency apache-airflow>=2.4.0.

This is a very legitimate case and there is no other way to express those dependencies to get them 'required'. You can make them into optional extras but not into requirements.

If you just patch the packages and remove one if those  dependencies one of the bad things happen:

- if you remove SMTP from apache-airflow, then someone installing just airflow will not have SMTP installed (which is bad)

- if you remove apache-airflow >= 2.4.0 from SMTP,  you will be able to install new provider with old version of airflow - which won't work.

So if you want to patch packages you have to remember the dependencies and apply them differently to serve similar cases  

-

groodt commented:
Yes. Thanks for that. And thats precisely the same approach we use at my workplace. Except, rather than trying to overload the concept of `annotations` or create a custom DSL around what we can / cannot do to the contents of external dependencies, we simply expose a `patches` argument (with similar semantics to http_file) which lets us patch whatever we like.

OliverFM commented:
What is blocking further progress on this PR? Since the solution in [this comment](https://github.com/bazelbuild/rules_python/pull/1166#issuecomment-1641878169) seems to be the defacto standard, but, unfortunately this patch doesn't work with the latest version of rules_python.
It would be great if we could get this merged and stop depending on patches

nkey0 commented:
> What is blocking further progress on this PR? Since the solution in [this comment](https://github.com/bazelbuild/rules_python/pull/1166#issuecomment-1641878169) seems to be the defacto standard, but, unfortunately this patch doesn't work with the latest version of rules_python. It would be great if we could get this merged and stop depending on patches

```
diff --git python/pip_install/tools/wheel_installer/wheel_installer.py python/pip_install/tools/wheel_installer/wheel_installer.py
index c6c2961..b0f9f76 100644
--- python/pip_install/tools/wheel_installer/wheel_installer.py
+++ python/pip_install/tools/wheel_installer/wheel_installer.py
@@ -129,7 +129,11 @@ def _extract_wheel(
     # Packages may create dependency cycles when specifying optional-dependencies / 'extras'.
     # Example: github.com/google/etils/blob/a0b71032095db14acf6b33516bca6d885fe09e35/pyproject.toml#L32.
     self_edge_dep = set([whl.name])
-    whl_deps = sorted(whl.dependencies(extras_requested) - self_edge_dep)
+    deps = whl.dependencies(extras_requested)
+    if (whl.name == "triton") and ("torch" in deps):
+        deps.remove("torch")
+    
+    whl_deps = sorted(deps - self_edge_dep)
 
     with open(os.path.join(installation_dir, "metadata.json"), "w") as f:
         metadata = {

```


```
http_archive(
    name = "rules_python",
    patches = [
        "//tools/patches:circulardeps.patch",
    ],
    sha256 = "5868e73107a8e85d8f323806e60cad7283f34b32163ea6ff1020cf27abef6036",
    strip_prefix = "rules_python-0.25.0",
    url = "https://github.com/bazelbuild/rules_python/releases/download/0.25.0/rules_python-0.25.0.tar.gz",
)
```

You could use such simple option for now.

groodt commented:
We've got a few options in the works on landing some basic "patch" functionality to the pip rules (not annotations). Working out the details of the API and how to integrate it with bzlmod etc.

In the short-term, the workaround above is the best short-term workaround for anyone running into cycles that they need to support and can't fix upstream or in the short-term.

arrdem commented:
Baselining question as I'm starting to overhaul this patch to actually apply again -- do we like the `cluster` terminology? Or do we want to call them `cycles` or something else little less ... artful.

groodt commented:
@arrdem We're likely going in this direction fwiw: https://github.com/bazelbuild/rules_python/pull/1393

arrdem commented:
I'm aware of the patch, although I don't see it as a replacement for this work. Patching out the dependency cycle and implementing the required closure yourself requires your users and tooling being aware of the cycle rather than oblivious to it, which is the interface PyPi provides.

groodt commented:
To confirm, it's not PyPI providing any interface. It's the fact that python installation into site-packages layouts doesn't check for cycles. Bazel does care about cycles and also doesn't use a site-packages layout, so patching is necessary anyway.

To reduce complexity and maintenance burden on the team, we've decided to provide first-class support for patching, which addresses cycles (and a host of other useful things), than introducing further complexity.

alexeagle commented:
I think our experience from rules_js may be useful here. We started from a google3 approach of patching the `require` algorithm in the node interpreter, a bit like how rules_python creates the runtime stub with a very long sys.path. This caused incompatibility with third-party packages and tools. The Google mantra at the time was "tools aren't expected to be compatible with Bazel, anyway Googlers always re-write them".

Then we took a different philosophy: why are we patching the runtime? What if we laid out the packages on disk exactly the way a standard package manager does? And by closely mirroring `pnpm` (porting parts of it to Starlark) we achieved this. Now, you can use an idiomatic JavaScript build tool for some framework (take Vue for example: https://github.com/bazelbuild/examples/blob/main/frontend/vue/BUILD.bazel) WITHOUT any patching or accommodation for Bazel.

> Bazel does care about cycles and also doesn't use a site-packages layout

Not using a site-packages layout is the same mistake we made with rules_nodejs, I believe. This is a choice, not a requirement. In aspect-build/rules_py we're building the layer on top of rules_python that does lay out `site-packages` in a virtualenv exactly like `pip` does, so tools are expected to work with Bazel unchanged.

For cycles, you're right that Bazel doesn't care about them - it just can't express them in the dependency graph, so cycles have to be within a single node in the graph. I think our philosophical goal should be that we find the right way for Bazel to accommodate the existing tools and not suggest to users that they have to change them, or that they should have been "written with Bazel in mind".

siddharthab commented:
I really agree with what Alex said. Pushing the problem back to users to avoid complexity in the tools might be OK in the short term or in certain cases, but these are foundational things and maintaining higher fidelity between Bazel and not-Bazel contexts makes the user onboarding and interchangeability story much simpler. One could argue that Bazel was not built with interpreted languages in mind (it's an action dependency graph), but there are many things we can still do.

As an anecdote of my own, in [rules_r](https://github.com/grailbio/rules_r), I also used to construct a long library search path composed of individual package locations in the Bazel execution root, thinking it's closer to Bazel philosophy. But we would occasionally hit failures when this path would exceed character limits imposed on it. Such failure modes are not hit outside of the Bazel context and took the users some time to troubleshoot. Switching to a flat symlink structure in a new directory took less than one hour of feature development for me, and has been a much smoother experience for everyone.

groodt commented:
Yes, precisely. A more natural layout for third-party packages is a direction we wish to move in. This PR wouldn't be moving things in a this direction. It would be introducing new concepts and user-facing complexity to support, when the complexity isn't with the cyclic dependencies per se, the complexity is with the historical decisions made by these rules and the context within which those decisions were made.

We have had some discussions about how to make the layout and `PYTHONPATH` a more natural fit for third-party packages. However, as I'm sure you realise, it's a big change that will take a while to land. It should hopefully be easier once [the Starlark implementation of the rules is the default](https://github.com/bazelbuild/rules_python/issues/1069)  because then the rules can be free to move without bazelbuild changes.

The need for `patch` semantics *will not* be removed when the layout is made to be more natural. It might be diminished slightly for the rare cases of cycles, but patching isn't only a solution to cycles. That's only one area where it is very useful.

That then naturally leads to the following:
1. This PR solves a cyclic dependency problem by asking the user to take manual steps to annotate a new abstraction "requirement groups". This is a non-standard abstraction in python packaging and pushes the problem back to the user in a similar way that patching would. It doesn't improve the layout of packages at installation time.
2. `patch` semantics are a necessary capability that can solve a wide variety of general issues with third-party dependencies. It can also be used to solve the cyclic dependency issues mentioned in the PR (i.e. https://github.com/bazelbuild/rules_python/issues/1076, https://github.com/bazelbuild/rules_python/issues/1188). It doesn't introduce additional abstractions (requirement groups) and it would still be required if there was an improved layout of packages at installation time.

|                    | Problems Addressed                                                                                                                          | Users Manually Address Cycles | Maintenance / Complexity                                                                 | Improves layout of third-party dependencies            |
|--------------------|---------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------|------------------------------------------------------------------------------------------|--------------------------------------------------------|
| requirement groups (this PR) | 1. cyclic dependencies                                                                                                                      | Yes                           | A new single-purpose abstraction with additional complexity and maintenance burden       | No, it changes the layout to be even more non-standard |
| patch semantics    | 1. cyclic dependencies 2. invalid or corrupt wheel RECORD 3. absolute paths to dynamic libs 4. code that loads outside bazel sandbox 5. etc | Yes                           | A multi-purpose abstraction, built into bazel with low complexity and maintenance burden | N/A patching isn't setting out to change layouts       |
|                    |                                                                                                                                             |                               |                                                                                          |                                                        |

arrdem commented:
Okay. It took a bit of head standing but I'd call this take II drastically cleaner. The previous magic in the `requirement()` macro has been eliminated entirely, and the changes to how requirements are extracted during wheel installation made managing the deps to avoid cycles also drastically simpler.

arrdem commented:
@chrislovecnm @rickeylev Sorry this patch sat for so long, it's rebased functional and better tested than the original. I'd appreciate it if you'd take another review pass.

Also it looks like the windows `pip_parse` integration tests are persistently failing for unrelated reasons. If you've got any pointers on resolving this I'd appreciate them https://buildkite.com/bazel/rules-python-python/builds/6109#018b6548-a0bb-4d82-aeee-2e4ce480d10a

arrdem commented:
@groodt I concur that this pushes the burden of managing cycles back onto the repository owner in the same sense patching cycles does. I don't see a way to get away from that.

I suggest that prioritizing the simplicitly of `rules_python`'s implementation over the ergonomics available both to repo owners and ultimately dependency users is not the right tradeoff. We should be absorbing complexity (within reason) here so that users get behavior they expect.

if a user (or tooling a user runs) says `requirement("apache-airflow")`, an `ImportError` in production because `@pypi_apache_airflow` doesn't depend on `@pypi_apache_airflow_providers_ftp` because the repo owner "patched the cycle out" would be surprising and undesirable. If a user typed `pip install apache-airflow` into a virtualenv or included it as part of a `pex` requirement set, they'd get `apache-airflow-providers-ftp` despite the cycle and without any intervention.

Similarly as a repository owner, authoring and maintaining patches is actually considerably higher effort than saying "these deps run together, make it so". Thanks to the intervening changes to the `whl_library` implementation, making groups work is now considerably simpler than it was before since we don't need to hack the installer -- just add a tiny bit of filtering and some `alias()` rules directly to the `BUILD.bazel` macro machinery. The complexity of this implementation is not that significant, and the convenience of not having to keep up patches especially for package groups like `apache-airflow` where many packages need to be "fixed" is high.

With regards to making the repository layout more nonstandard, I don't see that as an essential part of this change at all. The only essential part here is that we define groups and create alias and wrapper rules which allow us to implement them. This mechanism could easily be made to function if `rules_python` were to only create one repository and one `BUILD.bazel` in the future. If you feel strongly that the `@pip` repository should be used for these rules, I'd be happy to make that change. Today we seem to be leaning towards special-purpose repositories so this patch tries to fit with that pattern.

aignas commented on examples/pip_parse/WORKSPACE:
```diff
@@ -41,6 +41,19 @@ pip_parse(
 
+    # Requirement groups allow Bazel to tolerate PyPi cycles by putting dependencies
+    # which are known to form cycles into groups together.
+    requirement_groups = {
```
I am not sure the `requirement_groups` terminology is the best. We could call them what they are - dependency-cycles but I am not sure what the best term could be.

arrdem replied:
I think `groups` is less opaque than the original `clusters`, but it'd be easy enough to tweak things to reference `cycles` explicitly.

aignas replied:
What about:
```suggestion
    # The documentation for this could say that when 'pip' dependency cycles are
    # seen, they can use 'resolve_cycles' attribute to work-around this. In the
    # long term we should not need this argument.
    resolve_cycles = [
        [
            "sphinx",
            "sphincontrib-qthelp",
            "sphinxcontrib-htmlhelp"
            ...
        ],
    ]
```

The key argument is technically not needed because we could always get the first element of the `deps` list and then use that to construct a group name. For example, for this case we could have `sphinx` being used automatically.

aignas commented on examples/pip_parse/WORKSPACE:
```diff
@@ -41,6 +41,19 @@ pip_parse(
+    # Requirement groups allow Bazel to tolerate PyPi cycles by putting dependencies
+    # which are known to form cycles into groups together.
+    requirement_groups = {
+        "airflow": [
+            "apache-airflow",
+            "apache-airflow-providers-common-sql",
+            "apache-airflow-providers-ftp",
+            "apache-airflow-providers-http",
+            "apache-airflow-providers-imap",
+            "apache-airflow-providers-sqlite",
+        ],
```
I think using `airflow` as an example for this is not great. The reason why the providers are linking back to a particular version of `airflow` is for compatibility reasons. This dependency is used when doing version resolution (i.e. a particular version of an `apache-airflow-provider` is incompatible with older `airflow` versions) and they don't have actual import cycles, so it is perfectly fine to cut the dependency cycle by patching out the `airflow` dependency in the providers. 

That said, lumping them all into one group also makes sense conceptually, because even after patching things out, we will always have the list of packages present together.

aignas commented on examples/pip_parse/requirements.in:
```diff
@@ -1,3 +1,5 @@
 requests~=2.25.1
 s3cmd~=2.1.0
 yamllint~=1.28.0
+apache-airflow
+apache-airflow-providers-sqlite
```
Maybe we could use the following cycle as an example instead?
```
.-> @docs_deps_sphinx//:pkg
|   @docs_deps_sphinxcontrib_serializinghtml//:pkg
`-- @docs_deps_sphinx//:pkg
```

It could be much more lightweight. Some of `airflow` deps do not have wheels and having a smaller example may be better.

arrdem replied:
I'll give this a shot. I have a sneaking suspicion that the windows build problems are due to wheels / builds somehow so going for a purelib cycle would make for a better example.

aignas commented on examples/pip_parse_vendored/requirements.bzl:
```diff
@@ -5,10 +5,12 @@ from //:requirements.txt
 
 all_requirements = ["@pip_certifi//:pkg", "@pip_charset_normalizer//:pkg", "@pip_idna//:pkg", "@pip_requests//:pkg", "@pip_urllib3//:pkg"]
 
+all_requirement_groups = {}
```
Ideally we would call `pip install report --dry-run` during creating `requirements.bzl` file and detect these `groups` programatically. What do you think?

arrdem replied:
I'd love that honestly. In my ideal world this would Just Work (TM) magically without users having to think about cycles. If we can run that command maybe during the initial `pip_parse` invocation we could definitely try to drive groups automatically.

aignas commented on python/pip_install/pip_repository_requirements.bzl.tmpl:
```diff
@@ -8,6 +8,8 @@ from %%REQUIREMENTS_LOCK%%
 
 all_requirements = %%ALL_REQUIREMENTS%%
 
+all_requirement_groups = %%ALL_REQUIREMENT_GROUPS%%
```
Can we avoid using a global variable here? Why do we need to expose this value to our users?

arrdem replied:
If we're doing the groups dance correctly this should be totally transparent -- happy to make this private.

aignas commented on python/pip_install/pip_repository.bzl:
```diff
@@ -331,6 +337,7 @@ def _pip_repository_impl(rctx):
             "@{}//{}".format(rctx.attr.name, p) if rctx.attr.incompatible_generate_aliases else "@{}_{}//:pkg".format(rctx.attr.name, p)
```
Shouldn't we remove the files that are in `groups` from the `requirements` array? This would prevent users getting errors when generating the `gazelle` manifest.

arrdem replied:
We should be. This is an oversight on my part.

arrdem replied:
So something's not quite right over there. We don't need to filter this list, because we're using the same alias machinery for the `:whl` targets that we are for the `:pkg` targets. So to take your sphinx example, you get `@pip_sphinx//:whl` which an alias for `@pip__groups//:sphinx_whl`, which wraps up all the `:_whl` filegroups which lack inter-group dependencies. That this list is taking multiple dependencies (via aliases) on the group rules works out just fine.

Using that machinery --

```shellsession
$ bazel cquery --output=files 'deps(@pip//sphinx:whl)'
INFO: Analyzed target @pip//sphinx:whl (1 packages loaded, 1 target configured).
INFO: Found 1 target...
external/pip_sphinxcontrib_applehelp/sphinxcontrib_applehelp-1.0.7-py3-none-any.whl
external/pip_sphinx/sphinx-7.2.6-py3-none-any.whl
external/pip_sphinxcontrib_devhelp/sphinxcontrib_devhelp-1.0.5-py3-none-any.whl
external/pip_sphinxcontrib_serializinghtml/sphinxcontrib_serializinghtml-1.1.9-py3-none-any.whl
external/pip_sphinxcontrib_htmlhelp/sphinxcontrib_htmlhelp-2.0.4-py3-none-any.whl
external/pip_sphinxcontrib_qthelp/sphinxcontrib_qthelp-1.0.6-py3-none-any.whl
external/pip_packaging/packaging-23.2-py3-none-any.whl
external/pip_importlib_metadata/importlib_metadata-5.2.0-py3-none-any.whl
external/pip_snowballstemmer/snowballstemmer-2.2.0-py2.py3-none-any.whl
external/pip_alabaster/alabaster-0.7.13-py3-none-any.whl
external/pip_requests/requests-2.31.0-py3-none-any.whl
external/pip_pygments/Pygments-2.16.1-py3-none-any.whl
external/pip_sphinxcontrib_jsmath/sphinxcontrib_jsmath-1.0.1-py2.py3-none-any.whl
external/pip_docutils/docutils-0.20.1-py3-none-any.whl
external/pip_jinja2/Jinja2-3.1.2-py3-none-any.whl
external/pip_imagesize/imagesize-1.4.1-py2.py3-none-any.whl
external/pip_babel/Babel-2.13.1-py3-none-any.whl
external/pip_idna/idna-3.4-py3-none-any.whl
external/pip_markupsafe/MarkupSafe-2.1.1-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
external/pip_charset_normalizer/charset_normalizer-3.3.1-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
external/pip_zipp/zipp-3.11.0-py3-none-any.whl
external/pip_urllib3/urllib3-2.0.7-py3-none-any.whl
external/pip_certifi/certifi-2023.7.22-py3-none-any.whl
INFO: Elapsed time: 0.252s, Critical Path: 0.00s
INFO: 0 processes.
INFO: Build completed successfully, 0 total actions
```

https://gist.github.com/arrdem/5878f481c86b046df339af494bd1be7b

---

The bad news is that for reasons that aren't obvious to me the `//:gazelle_python_manifest.update` rule isn't picking up all the component wheels for some reason despite `bazel query` giving the desired results. Do we want to be referencing the concrete `:_whl` targets here if the Gazelle indexing machinery doesn't compute a closure and traverse filegroup deps fully?

arrdem replied:
Ah. The manifest generator rule doesn't traverse `data=[]` in filegroups. Easy fix.

arrdem replied:
Gazelle manifest generation works fine now

aignas reviewed (changes requested):
I am not sure if this would work with `gazelle`. Right now the gazelle manifest generation pulls all wheels and it uses `all_whl_requirements` array. The `whl` filegroups will have circular dependencies, because here we are not patching them and will cause the operation to fail.

This means that so far `patching` out the well known cycles and `e.g.` storing them in the `rules_python` would be a better solution for now.

I am also interested in knowing how we could do `pip install --report - --dry-run -r requirements.txt` to detect the dependency cycles and construct the `requirement_groups` automatically.

After a second thought, I am not opposed to the idea of doing this under the hood as contracting the nodes forming a cycle in a graph could make us a directed acyclic graph reliably, so this idea in general is promising. Since `bzlmod` will be enabled by default with `7.0` it would be great to merge `bzlmod` support in the same PR. 

@arrdem, what do you think?

groodt commented:
Agree with @aignas 

Thanks for making a simpler implementation here @arrdem. It looks a lot less intimidating than the original proposals to maintain. However, I still think we'd need this to work with bzlmod at a minimum before accepting it.

Independently, I think I'm also going to start an issue to start recording the instances of circular deps that are being encountered in the wild. It might help to highlight if we need to change the way that we approach third_party deps entirely. Or if it's only happening for a few rare packages.

arrdem commented:
@aignas 
> I am not sure if this would work with gazelle. Right now the gazelle manifest generation pulls all wheels and it uses all_whl_requirements array. The whl filegroups will have circular dependencies, because here we are not patching them and will cause the operation to fail.

Let me try and confirm that. I'm deliberately applying the same group strategy both to libs and to whls, we should be able to make that work. My use case for this patch ultimately is enabling transparent dependency automation through my own pipeline.

@groodt Thanks for taking another look! I started poking into the bzlmod side of the house last night and I don't see a reason why we couldn't make this work. Will take a more concerted pass.

aignas commented on examples/pip_parse_vendored/requirements.bzl:
```diff
@@ -5,10 +5,12 @@ from //:requirements.txt
 
 all_requirements = ["@pip_certifi//:pkg", "@pip_charset_normalizer//:pkg", "@pip_idna//:pkg", "@pip_requests//:pkg", "@pip_urllib3//:pkg"]
 
+_all_requirement_groups = {}
```
please move this global variable to the inside of `install_deps`

arrdem commented:
I'd suggest that we leave automating groups to a future PR. It isn't super trivial because you have to run the dry-run install, parse the resulting manifest, deal with platform constraints and extras, and then run a connected components implementation on the resulting graph. It also sorta undoes all the hard work of `requirements_parser.bzl` by putting a Python interpreter phase back into setup. It's definitely doable, I just don't think it'll be dead simple.

aignas commented on python/pip_install/private/generate_whl_library_build_bazel.bzl:
```diff
@@ -154,28 +176,37 @@ def generate_whl_library_build_bazel(
+
     lib_dependencies = [
-        "@" + repo_prefix + normalize_name(d) + "//:" + _PY_LIBRARY_LABEL
+        "@" + repo_prefix + normalize_name(d) + "//:" + PY_LIBRARY_PUBLIC_LABEL
         for d in dependencies
+        if d not in group_deps
     ]
     whl_file_deps = [
-        "@" + repo_prefix + normalize_name(d) + "//:" + _WHEEL_FILE_LABEL
+        "@" + repo_prefix + normalize_name(d) + "//:" + WHEEL_FILE_PUBLIC_LABEL
         for d in dependencies
+        if d not in group_deps
     ]
```
nit: having `non_group_deps = [d for d in dependencies if d not in group_deps]` and then using that to generated `lib_dependencies` and `whl_file_deps` would be great!

aignas commented on python/pip_install/private/labels.bzl:
```diff
@@ -0,0 +1,24 @@
+# Copyright 2023 The Bazel Authors. All rights reserved.
```
Please move this to somewhere under `python/private`. In the long term we would like to have all private files there. I'll create a small PR to do that in parallel, so that we could agree on the good location.

aignas commented on tests/pip_install/whl_library/generate_build_bazel_tests.bzl:
```diff
@@ -219,6 +249,74 @@ py_binary(
+    actual = generate_whl_library_build_bazel(
+        repo_prefix = "pypi_",
+        whl_name = "foo.whl",
+        dependencies = ["foo", "bar-baz", "qux"],
```
The `whl_installer` is going to remove `self` from the dependency list, so `"foo"` is likely to never be present in the `dependencies` list here.

arrdem replied:
Yeah this is just copypasta from the previous test https://github.com/bazelbuild/rules_python/pull/1166/files#diff-864e4b8f44994e69a2522a561ca85c7721aa1bd869ef9223daf9996d921cdb24R242. Happy to clean both or just this instance up.

aignas commented on tests/pip_install/whl_library/generate_build_bazel_tests.bzl:
```diff
@@ -219,6 +249,74 @@ py_binary(
+)
+
+filegroup(
+    name = "_whl",
```
Why do we need this label here?

arrdem replied:
In this example, this is the filegroup which `@pypi__groups//:qux_whl` depends on, same as `@pypi__groups//:qux_pkg` depends on `:_pkg`. See https://github.com/bazelbuild/rules_python/pull/1166/files#diff-ecdc677552e2995c1c31461205f3ce1d875bcafc15453b568644f77c3d37d7d8R29 for the tests on that side.

arrdem commented:
If we want to do something like `resolve_cycles=List[List[Requirement]]`, my immediate concern is that I don't think "naming by first listed" is the right choice for naming since cycles are really unordered. Naming by first in sort order, naming by a hash of the sorted components, ordinal naming or just letting the user provide a name seem like better choices.

My reason for letting users provide names was so there wasn't some surprisingly named element in the build graph.

arrdem commented:
FTR:
- Got `gazelle fix` working
- Got cycle breaking under bzlmod working

aignas commented on python/pip_install/private/generate_group_library_build_bazel.bzl:
```diff
@@ -0,0 +1,105 @@
+        A complete BUILD file as a string
+    """
+
+    content = [_PRELUDE]
+
+    for group_name, group_members in groups.items():
+        content.append(_generate_group_libraries(repo_prefix, group_name, group_members))
```
nit: I am a fan of list comprehensions, so let's see how it looks like.

```suggestion
    content = [_PRELUDE] + [
        _generate_group_libraries(repo_prefix, group_name, group_members)
        for group_name, group_members in groups.items()
    ]
```

aignas commented on python/pip_install/private/generate_whl_library_build_bazel.bzl:
```diff
@@ -83,6 +87,16 @@ py_library(
     tags = {tags},
 )
+
+alias(
+   name = "{py_library_public_label}",
+   actual = "{py_library_actual_label}",
+)
```
The new alias should be public, but the `impl` targets should be IMHO private. So an explicit visibility setting would be good.

aignas commented on python/pip_install/private/generate_group_library_build_bazel.bzl:
```diff
@@ -0,0 +1,105 @@
+          which make up the group.
+    """
+
+    lib_dependencies = [
+        "@" + repo_prefix + normalize_name(d) + "//:" + PY_LIBRARY_IMPL_LABEL
+        for d in group_members
+    ]
+    whl_file_deps = [
+        "@" + repo_prefix + normalize_name(d) + "//:" + WHEEL_FILE_IMPL_LABEL
+        for d in group_members
+    ]
+    visibility = [
+        "@" + repo_prefix + normalize_name(d) + "//:__pkg__"
+        for d in group_members
+    ]
```
Thanks for setting explicit visibility.

aignas replied:
nit: This may be easier to read:
```suggestion
    repo_name = "@%s%s" % (repo_prefix, normalize_name(d))
    lib_dependencies = [
        "@%s//:%s" % (repo_name, PY_LIBRARY_IMPL_LABEL)
        for d in group_members
    ]
    whl_file_deps = [
        "@%s//:%s" % (repo_name, WHEEL_FILE_IMPL_LABEL)
        for d in group_members
    ]
    visibility = [
        "@%s//:%s" % (repo_name, "//:__pkg__")
        for d in group_members
    ]
```

aignas commented on python/pip_install/pip_repository_requirements.bzl.tmpl:
```diff
@@ -43,12 +43,40 @@ def _get_annotation(requirement):
+
+    all_requirement_groups = %%ALL_REQUIREMENT_GROUPS%%
+
     whl_config = dict(_config)
     whl_config.update(whl_library_kwargs)
```
nit: moving this to just before the for loop would make things a little easier to read. Right now we have a lot of things going on in the code here.

aignas commented on python/pip_install/pip_repository.bzl:
```diff
@@ -694,6 +712,13 @@ whl_library_attrs = {
         allow_files = True,
     ),
+    "group_deps": attr.string_list(
+        doc = "List of requirements to skip in order to break the cycles within a dependency group.",
```
```suggestion
        doc = "List of dependencies to skip in order to break the cycles within a dependency group.",
```

aignas commented on python/pip_install/pip_repository.bzl:
```diff
@@ -694,6 +712,13 @@ whl_library_attrs = {
+        default = [],
+    ),
+    "group_name": attr.string(
+        doc = "Name of the group, if any.",
```
```suggestion
        doc = "Name of the dependency cycle this belongs to.",
        mandatory = False,
```

aignas commented on python/pip_install/pip_repository.bzl:
```diff
@@ -424,6 +431,15 @@ python_interpreter. An example value: "@python3_x86_64-unknown-linux-gnu//:pytho
+    "requirement_groups": attr.string_list_dict(
+        default = {},
+        doc = """\
+A mapping of requirements which form dependency cycles into groups.
+
+Groups of packages will be wrapped so that if a dependency is taken on a member of the cycle the rest of the cycle will
+be correctly included as transitive dependencies.
```
It would be nice to improve the docs to be more user friendly.

```suggestion
A mapping of dependency cycle names to the list of packages that form a cycle.

This attribute allows `rules_python` to work around cycles in Python dependency graph. If you see a dependency cycle like
```
EXAMPLE error
```

Then you can include the following key, value pair to resolve the error:
```
EXAMPLE
```

Note that the key of this dictionary is arbitrary
and the recommendation is to use a value that is easy for humans to debug when looking at dependency graphs.

Adding a group of packages here will ensure that
all of the packages in the dependency cycle are included in the transitive dependency list when any of them is used as dependencies in your code or third party Python libraries.
```

aignas commented on python/pip_install/pip_repository.bzl:
```diff
@@ -424,6 +431,15 @@ python_interpreter. An example value: "@python3_x86_64-unknown-linux-gnu//:pytho
 Prefix for the generated packages will be of the form `@<prefix><sanitized-package-name>//...`
 """,
     ),
+    "requirement_groups": attr.string_list_dict(
```
What about having `circular_deps` name to this? This is the best I can think off right now.

```suggestion
    "circular_deps": attr.string_list_dict(
```

aignas commented on examples/build_file_generation/requirements.in:
```diff
@@ -1 +1,4 @@
 flask
+sphinx
+sphinxcontrib-serializinghtml
+colorama  # platform-conditional dependency of sphinx
```
Why do we need it here? I would assume that locking the requirements on `windows` would work here.

arrdem replied:
Should have cleaned this up.

aignas commented on examples/pip_parse/BUILD.bazel:
```diff
@@ -38,6 +39,8 @@ py_binary(
     srcs = ["main.py"],
     deps = [
         "@pypi_requests//:pkg",
+        requirement("sphinx"),
+        "@pypi_sphinxcontrib_serializinghtml//:pkg",
```
We are showing here that both are possible?

aignas reviewed (commented):
Nice work, the trick of re-direction here works really nicely. From my view the following are needed:
* Finalize naming.
* Add a CHANGELOG entry.
* Add a small design note for the `group_library` so that future maintainers can reason about the code more easily.

arrdem commented:
Okay. Got @aignas's comments sorted out, build is green, docs and changelog are updated.

One thing I note is that this approach to repairing dependency cycles promotes optional dependencies to strict dependencies. For instance `apache-airflow-providers-postgres` is cyclic with `apache-airflow`, which means it has to become part of the same cycle and thus a dep of targets which may never use Postgres. Maybe this makes sense given how `rules_python` handles requirement features, but it's also a case where to @groodt's previous commentary maybe we want to just remove one leg of the cycle via patching or something else. I don't think it really makes sense to introduce special behavior for breaking let's call them weak cycles, but I could see cases where this behavior could potentially produce build issues.

aignas reviewed (approved):
Thanks, given that this approach is super fresh, what do you think about adding `experimental_` prefix to the newly added parameter? It would allow us to make changes in the future releases more freely. Since `bzlmod` is all `beta` we might be fine there as is, but the `legacy` `pip_repository` is not a `beta` API.

@rickeylev, what do you think about parameter naming it?

This looks good enough to me to merge and then improve it later on so that we can benefit from this new machinery whilst we get more experience with it and see how we can make improvements further. I'll defer to other to have a look now.

arrdem commented on docs/sphinx/pypi-dependencies.md:
````diff
@@ -130,6 +130,70 @@ Any 'extras' specified in the requirements lock file will be automatically added
+```
+
+`pip_parse` supports fixing multiple cycles simultaneously, however cycles must
+be distinct. `apache-airflow` for instance has dependency cycles with a number
````
Sure thing. Had to think about whether this approach could support support this or not and came down on the side of not hence the new validation pass in `pip_repository.bzl`.

arrdem commented on examples/bzlmod/MODULE.bazel:
```diff
@@ -90,6 +90,16 @@ use_repo(pip, "whl_mods_hub")
 pip.parse(
     hub_name = "pip",
     python_version = "3.9",
+    requirement_cycles = {
```
I think that makes sense for now. Struggling a little bit with the naming and what it'd look like to have a way to handle a cycle this way by merging or the alternate strategy of breaking it, and whether it'd make sense to have separate `merge_requirement_cycles` and `break_requirement_cycles` behavior explicitly.

arrdem commented on python/pip_install/pip_repository.bzl:
```diff
@@ -278,8 +280,38 @@ def _pip_repository_impl(rctx):
+            right_group = requirement_cycles.keys()[1 + i + j]
+            for d in left_deps:
+                if d in right_deps:
+                    fail("Error: Requirement %s is repeated between cycles %s and %s, they must be merged." % (d, left_group, right_group))
```
No that I can think of because the repeated dependency becomes ambiguous. If a user specifies `@pypi//foo:pkg` which is a member of two separate cycles, how do we determine which cycle the user intended to pull in? I think that only makes sense in a world where we have a way to do explicit cycle breaking (via patching or a separate flag) rather than cycle consolidation consolidation which is the core of this approach.

arrdem replied:
Mostly thinking out loud here but the only reasonable thing we could do would be to transparently merge cycles with shared requirements on behalf of the user. It'd be nice to be a bit more declarative but Bazel generally errs on the side of explicit so it'd be out of place.

arrdem commented on docs/sphinx/pypi-dependencies.md:
````diff
@@ -130,6 +130,70 @@ Any 'extras' specified in the requirements lock file will be automatically added
+)
+```
+
+`pip_parse` supports fixing multiple cycles simultaneously, however cycles must
+be distinct. `apache-airflow` for instance has dependency cycles with a number
+of its optional dependencies, which means those optional dependencies must all
+be a part of the `airflow` cycle. For instance --
+
+```
+pip_parse(
+  ...
+  requirement_cycles = {
+    "airflow": [
+      "apache-airflow",
+      "apache-airflow-providers-common-sql",
+      "apache-airflow-providers-postgres",
+      "apache-airflow-providers-sqlite",
+    ]
+  }
+)
+```
````
Added some more notes on this.

arrdem commented:
@rickeylev would you please take a look when you get the chance

rickeylev commented:
This is some pretty solid work. Thanks, @arrdem!
